File: windows/Hinge.Core/TrustStore.cs
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace Hinge.Core;

public class TrustedDevice
{
    [JsonPropertyName("deviceId")]
    public string DeviceId { get; set; } = string.Empty;

    [JsonPropertyName("name")]
    public string Name { get; set; } = string.Empty;

    [JsonPropertyName("publicKey")]
    public string PublicKey { get; set; } = string.Empty;

    [JsonPropertyName("pairedAt")]
    public long PairedAt { get; set; } = DateTimeOffset.UtcNow.ToUnixTimeSeconds();

    [JsonPropertyName("lastSeen")]
    public long LastSeen { get; set; } = DateTimeOffset.UtcNow.ToUnixTimeSeconds();

    [JsonPropertyName("trustState")]
    public TrustState TrustState { get; set; } = TrustState.Trusted;
}

public class TrustStore
{
    private readonly string _storagePath;
    private readonly Dictionary<string, TrustedDevice> _trustedDevices = new();
    private readonly object _lock = new();

    public TrustStore(string? storagePath = null)
    {
        if (string.IsNullOrEmpty(storagePath))
        {
            string appData = Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData);
            string suiteDir = Path.Combine(appData, "Hinge");
            Directory.CreateDirectory(suiteDir);
            _storagePath = Path.Combine(suiteDir, "trust_store.json");
        }
        else
        {
            _storagePath = storagePath;
        }

        Load();
    }

    public bool IsTrusted(string deviceId)
    {
        lock (_lock)
        {
            return _trustedDevices.TryGetValue(deviceId, out var dev) && dev.TrustState == TrustState.Trusted;
        }
    }

    public TrustedDevice? GetDevice(string deviceId)
    {
        lock (_lock)
        {
            return _trustedDevices.TryGetValue(deviceId, out var dev) ? dev : null;
        }
    }

    public IReadOnlyList<TrustedDevice> GetAllTrustedDevices()
    {
        lock (_lock)
        {
            return _trustedDevices.Values.ToList();
        }
    }

    public void AddOrUpdate(TrustedDevice device)
    {
        lock (_lock)
        {
            _trustedDevices[device.DeviceId] = device;
            Save();
        }
    }

    public bool Revoke(string deviceId)
    {
        lock (_lock)
        {
            if (_trustedDevices.Remove(deviceId))
            {
                Save();
                return true;
            }
            return false;
        }
    }

    private void Load()
    {
        lock (_lock)
        {
            if (!File.Exists(_storagePath)) return;
            try
            {
                string json = File.ReadAllText(_storagePath);
                var list = JsonSerializer.Deserialize<List<TrustedDevice>>(json);
                if (list != null)
                {
                    _trustedDevices.Clear();
                    foreach (var item in list)
                    {
                        _trustedDevices[item.DeviceId] = item;
                    }
                }
            }
            catch
            {
                // Fallback for corrupt file
            }
        }
    }

    private void Save()
    {
        lock (_lock)
        {
            try
            {
                string json = JsonSerializer.Serialize(_trustedDevices.Values.ToList(), new JsonSerializerOptions { WriteIndented = true });
                File.WriteAllText(_storagePath, json);
            }
            catch
            {
                // Fallback
            }
        }
    }
}

File: android/lib/core/trust_store.dart
import 'dart:convert';
import 'dart:io';

import 'device_model.dart';

class TrustedDevice {
  final String deviceId;
  final String name;
  final String publicKey;
  final int pairedAt;
  final int lastSeen;
  final DeviceTrustState trustState;

  const TrustedDevice({
    required this.deviceId,
    required this.name,
    this.publicKey = '',
    required this.pairedAt,
    required this.lastSeen,
    this.trustState = DeviceTrustState.trusted,
  });

  Map<String, dynamic> toJson() => {
    'deviceId': deviceId,
    'name': name,
    'publicKey': publicKey,
    'pairedAt': pairedAt,
    'lastSeen': lastSeen,
    // Windows' former WinUI client used System.Text.Json's numeric enum
    // representation. Keep that format on the shared desktop path so an
    // existing trust_store.json remains readable by both clients.
    'trustState': Platform.isWindows ? trustState.index : trustState.name,
  };

  factory TrustedDevice.fromJson(Map<String, dynamic> json) {
    final rawTrustState = json['trustState'];
    final trustState = rawTrustState is num
        ? (rawTrustState.toInt() >= 0 &&
                  rawTrustState.toInt() < DeviceTrustState.values.length
              ? DeviceTrustState.values[rawTrustState.toInt()]
              : DeviceTrustState.trusted)
        : DeviceTrustState.values.firstWhere(
            (e) => e.name == rawTrustState,
            orElse: () => DeviceTrustState.trusted,
          );
    return TrustedDevice(
      deviceId: json['deviceId'] as String,
      name: json['name'] as String,
      publicKey: json['publicKey'] as String? ?? '',
      pairedAt: json['pairedAt'] as int? ?? 0,
      lastSeen: json['lastSeen'] as int? ?? 0,
      trustState: trustState,
    );
  }
}

class TrustStore {
  final String? _customPath;
  final Map<String, TrustedDevice> _trustedDevices = {};

  TrustStore([this._customPath]) {
    _load();
  }

  File _getStorageFile() {
    final custom = _customPath;
    if (custom != null) {
      return File(custom);
    }
    final home = Platform.isWindows
        ? (Platform.environment['LOCALAPPDATA'] ??
              Platform.environment['APPDATA'] ??
              Directory.systemTemp.path)
        : (Platform.environment['HOME'] ?? Directory.systemTemp.path);
    final dir = Directory(Platform.isWindows ? '$home/Hinge' : '$home/.hinge');
    if (!dir.existsSync()) {
      dir.createSync(recursive: true);
    }
    return File('${dir.path}/trust_store.json');
  }

  bool isTrusted(String deviceId) {
    final dev = _trustedDevices[deviceId];
    return dev != null && dev.trustState == DeviceTrustState.trusted;
  }

  TrustedDevice? getDevice(String deviceId) => _trustedDevices[deviceId];

  List<TrustedDevice> getAllTrustedDevices() => _trustedDevices.values.toList();

  void addOrUpdate(TrustedDevice device) {
    _trustedDevices[device.deviceId] = device;
    _save();
  }

  bool revoke(String deviceId) {
    if (_trustedDevices.remove(deviceId) != null) {
      _save();
      return true;
    }
    return false;
  }

  void _load() {
    final file = _getStorageFile();
    if (!file.existsSync()) return;
    try {
      final content = file.readAsStringSync();
      final list = jsonDecode(content) as List;
      _trustedDevices.clear();
      for (final item in list) {
        final dev = TrustedDevice.fromJson(item as Map<String, dynamic>);
        _trustedDevices[dev.deviceId] = dev;
      }
    } catch (_) {}
  }

  void _save() {
    final file = _getStorageFile();
    try {
      final list = _trustedDevices.values.map((d) => d.toJson()).toList();
      file.writeAsStringSync(jsonEncode(list));
    } catch (_) {}
  }
}
